File: euler43.py
# Найдите сумму всех пан-цифровых чисел из цифр от 0 до 9,
# обладающих данным свойством.

from datetime import datetime
from itertools import permutations
from typing import List, Optional


def eratosfen_lists(
    number: int,
    dimension: Optional[int] = None
) -> List[int]:
    """
    Решето Эратосфена: генерация списка простых чисел.

    Args:
        number: Верхняя граница диапазона.
        dimension: Размерность для начального числа.

    Returns:
        Список простых чисел.
    """
    start = 1 if dimension is None else dimension
    sieve = []
    for k in [
        x for x in range(10 ** (start - 1) + 1, number + 1)
        if x not in [
            i for sub in [
                list(range(2 * j, number + 1, j))
                for j in range(2, number // 2)
            ]
            for i in sub
        ]
    ]:
        sieve.append(k)
    return sieve


def euler43() -> None:
    """
    Решение задачи Эйлера №43.

    Найдите сумму всех пан-цифровых чисел из цифр от 0 до 9.
    """
    n = 10

    # 1 вариант
    start_time = datetime.now()
    pan_numbers = list(permutations(range(n), n))
    pan_numbers_sub, divisors = [], eratosfen_lists(18)
    for i in pan_numbers:
        if i[0] != 0:
            check_pan_number = [
                100 * i[y] + 10 * i[y + 1] + i[y + 2]
                for y in range(1, len(i) - 2)
            ]
            if all(
                check_pan_number[z] % divisors[z] == 0
                for z in range(len(check_pan_number))
            ):
                pan_numbers_sub.append(int(''.join(str(x) for x in i)))
    print(sum(pan_numbers_sub))
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    pass
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler43()

File: go/euler43.go
// euler43 находит сумму всех пан-цифровых чисел из цифр от 0 до 9.
//
// Решение задачи Эйлера №43.
package main

import (
	"fmt"
	"time"
)

func eratosfenLists(number int, dimension int) []int {
	start := 1
	if dimension > 0 {
		start = dimension
	}
	sieve := make([]int, 0)
	isComposite := make([]bool, number+1)
	for i := 2; i <= number/2; i++ {
		for j := 2 * i; j <= number; j += i {
			isComposite[j] = true
		}
	}
	for k := start; k <= number; k++ {
		if !isComposite[k] {
			sieve = append(sieve, k)
		}
	}
	return sieve
}

func euler43() {
	n := 10

	startTime := time.Now()
	panNumbers := generatePermutations10()
	panNumbersSub := make([]int, 0)
	divisors := []int{2, 3, 5, 7, 11, 13, 17}
	for _, i := range panNumbers {
		if i[0] != 0 {
			valid := true
			for y := 1; y <= len(i)-3; y++ {
				num := i[y]*100 + i[y+1]*10 + i[y+2]
				if num%divisors[y-1] != 0 {
					valid = false
					break
				}
			}
			if valid {
				num := 0
				for _, d := range i {
					num = num*10 + d
				}
				panNumbersSub = append(panNumbersSub, num)
			}
		}
	}
	sum := 0
	for _, v := range panNumbersSub {
		sum += v
	}
	fmt.Println(sum)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	fmt.Println(time.Since(startTime))
}

func generatePermutations10() [][]int {
	arr := []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
	return permute10(arr)
}

func permute10(arr []int) [][]int {
	result := [][]int{}
	permute10Helper(arr, 0, &result)
	return result
}

func permute10Helper(arr []int, i int, result *[][]int) {
	if i == len(arr) {
		tmp := make([]int, len(arr))
		copy(tmp, arr)
		*result = append(*result, tmp)
	} else {
		for j := i; j < len(arr); j++ {
			arr[i], arr[j] = arr[j], arr[i]
			permute10Helper(arr, i+1, result)
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
}

func main() {
	euler43()
}
